drivers/counter/counter_mcux_pit.c

```diff
@@ -179,41 +213,101 @@ static const struct counter_driver_api mcux_pit_driver_api = {
 	}
 
-DT_INST_FOREACH_STATUS_OKAY(COUNTER_MCUX_PIT_DEVICE)
+#define MCUX_PIT_CONFIG_FUNC(n)									\
+	static void mcux_pit_irq_config_func_##n(const struct device *dev)			\
+
+
+/* Creates a device for a channel (needed for counter API) */
+#define MCUX_PIT_CHANNEL_DEV_INIT(node, pit_inst)						\
+	DEVICE_DT_DEFINE(node, NULL, NULL,							\
+		(void *)									\
+		&mcux_pit_##pit_inst##_channel_datas[DT_REG_ADDR(node)],			\
+		&mcux_pit_##pit_inst##_config,							\
+		POST_KERNEL, CONFIG_COUNTER_INIT_PRIORITY,					\
+		&mcux_pit_driver_api);
+
+/* Creates a decleration for each pit channel */
+#define MCUX_PIT_CHANNEL_DECLARATIONS(node)							\
+	static struct mcux_pit_channel_data							\
+		mcux_pit_channel_data_##node;
+
+/* Initializes an element of the channel data pointer array */
+#define MCUX_PIT_INSERT_CHANNEL_INTO_ARRAY(node)						\
+	[DT_REG_ADDR(node)] =									\
+		&mcux_pit_channel_data_##node,
+
+#define MCUX_PIT_INSERT_CHANNEL_DEVICE_INTO_ARRAY(node)						\
+	[DT_REG_ADDR(node)] = DEVICE_DT_GET(node),
+
+
+#define COUNTER_MCUX_PIT_DEVICE_INIT(n)								\
+	/* Create Common ISR for 1 or multiple irqs */						\
+	static void mcux_pit_irq_config_func_##n(const struct device *dev)			\
+	{											\
+		LISTIFY(DT_NUM_IRQS(DT_DRV_INST(n)),						\
+			IRQ_CONFIG, (;), n)							\
+	};											\
+												\
+	/* Create channel declarations */							\
+	DT_INST_FOREACH_CHILD_STATUS_OKAY(n,							\
+		MCUX_PIT_CHANNEL_DECLARATIONS)							\
+												\
+	/* Array of channel devices */								\
+	static struct mcux_pit_channel_data *const						\
+		mcux_pit_##n##_channel_datas							\
+			[DT_INST_PROP(n, num_channels)] = {					\
+				DT_INST_FOREACH_CHILD_STATUS_OKAY(n,				\
+					MCUX_PIT_INSERT_CHANNEL_INTO_ARRAY)			\
+	};											\
+												\
+	/* forward declaration */								\
+	static const struct mcux_pit_config mcux_pit_##n##_config;				\
+	static void mcux_pit_irq_config_func_##n(const struct device *dev);			\
+												\
+	/* Create all the channel/counter devices */						\
+	DT_INST_FOREACH_CHILD_STATUS_OKAY_VARGS(n,						\
+		MCUX_PIT_CHANNEL_DEV_INIT, n)							\
+												\
+	/* This channel device array is needed by the module device ISR */			\
+	const struct device *const mcux_pit_##n##_channels					\
+					[DT_INST_PROP(n, num_channels)] = {			\
+					DT_INST_FOREACH_CHILD_STATUS_OKAY(n,			\
+					MCUX_PIT_INSERT_CHANNEL_DEVICE_INTO_ARRAY)		\
+	};											\
+												\
+	/* This config struct is shared by all the channels and parent device */		\
+	static const struct mcux_pit_config mcux_pit_##n##_config = {				\
+		.info = {									\
+			.max_top_value =							\
+				DT_INST_PROP(n, max_load_value),				\
+			.channels = 0,								\
+		},										\
+		.base = (PIT_Type *)DT_INST_REG_ADDR(n),					\
+		.irq_config_func = mcux_pit_irq_config_func_##n,				\
+		.num_channels = DT_INST_PROP(n, num_channels),					\
+		.pit_period = DT_INST_PROP(n, pit_period),					\
```

Comment:
this property and struct field (pit_period) needs to be removed

zephyr api says default counter top shall be max value, and this is not used anywhere

Reply:
I have removed the pit period field and will be updated the next push to this PR.